### PR TITLE
LPS-113562 Deprecate `deprecated.js`

### PR DESCRIPTION
**Pull request by wincent**
Just forwarding a copy of this here from https://github.com/wincent/liferay-portal/pull/306 because there is a chance you might get to it before I do. (I have been looking into [IFI-1716](https://issues.liferay.com/browse/IFI-1716) and not finished yet.)

---

`deprecated.js` has a [comment](https://github.com/liferay/liferay-portal/blob/9b4e4366ad5f71fc89f88fe351d8f0d96d9056cb/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/deprecated.js#L15) which mentions [LPS-2155](https://issues.liferay.com/browse/LPS-2155), in which it is said that `deprecated.js` was added to [`portal.properties`](https://github.com/liferay/liferay-portal/blob/4ae164e4f5859d26a26c7a2600dd40044aad0aa8/portal-impl/src/portal.properties) in order to allow deprecating JavaScript.

There's no mention of `deprecated.js` in this [`portal.properties`](https://github.com/liferay/liferay-portal/blob/4ae164e4f5859d26a26c7a2600dd40044aad0aa8/portal-impl/src/portal.properties) file, or at least not anymore because it was removed in [`60256e30d160`](https://github.com/liferay/liferay-portal/commit/60256e30d160). (Found with: `git log -G'\bdeprecated\.js\b' -p`)

Searching for usages of this file, with:

```
git grep -G "\b\/deprecated\.js\b"
```

shows this result:

```
modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/servlet/AUITopHeadResources.java: "/liferay/deprecated.js"
```

Here's a list of the usages of the constants and functions declared in `deprecated.js`

- `Liferay.Util.MAP_HTML_CHARS_ESCAPED`: *2 usages*

```
modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/bbcode/bbcode_parser.js: var entities = A.merge(Liferay.Util.MAP_HTML_CHARS_ESCAPED, {`
`modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/bbcode/bbcode_utils.js: var entities = A.merge(Liferay.Util.MAP_HTML_CHARS_ESCAPED, {
```

- `Liferay.Util.actsAsAspect`: 0 usages
- `Liferay.Util.addInputFocus`: 0 usages
- `Liferay.Util.addInputType`: 0 usages
- `Liferay.Util.camelize`: 0 usages
- `Liferay.Util.clamp`: 0 usages

- `Liferay.Util.escapeHTML`: *8 usages*

```
modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/input_asset_links/page.jsp: Liferay.Util.escapeHTML(`
`modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/input_asset_links/page.jsp: Liferay.Util.escapeHTML(`
`modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/input_asset_links/page.jsp: Liferay.Util.escapeHTML(`
`modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/input_localized.js: helpMessage = Liferay.Util.escapeHTML(helpMessage);`
`modules/apps/sharing/sharing-web/src/main/resources/META-INF/resources/dynamic_include/bottom.jsp: title: Liferay.Util.escapeHTML(title),`
`modules/dxp/apps/portal-workflow/portal-workflow-kaleo-forms-web/src/main/resources/META-INF/resources/admin/js/main.js: currentName = Liferay.Util.escapeHTML(currentName);`
`modules/dxp/apps/portal-workflow/portal-workflow-kaleo-forms-web/src/main/resources/META-INF/resources/admin/process/fields.jsp: Liferay.Util.escapeHTML(ddmStructureName)`
`modules/dxp/apps/portal-workflow/portal-workflow-kaleo-forms-web/src/main/resources/META-INF/resources/admin/process/workflow.jsp: title: Liferay.Util.escapeHTML(event.title),
```

- `Liferay.Util.isEditorPresent`: 0 usages
- `Liferay.Util.randomMinMax`: 0 usages
- `Liferay.Util.randomMinMax`: 0 usages
- `Liferay.Util.selectAndCopy`: 0 usages
- `Liferay.Util.setBox`: 0 usages
- `Liferay.Util.startsWith`: 0 usages
- `Liferay.Util.textareaTabs`: 0 usages
- `Liferay.Util.uncamelize`: 0 usages
- `Liferay.Util.unescapeHTML`: 0 usages
- `Liferay.Util._escapeHTML`: 0 usages
- `Liferay.Util._unescapeHTML`: 0 usages

Some other functions, are defined via [`Liferay.provide`](https://github.com/liferay/liferay-portal/blob/4ae164e4f5859d26a26c7a2600dd40044aad0aa8/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/dependency.js#L60-L152) on the `Liferay.Util` global object.

- `Liferay.Util.check`: 0 Usages
- `Liferay.Util.disableSelectBoxes`: 0 usages
- `Liferay.Util.disableTextareaTabs`: 0 usages
- `Liferay.Util.enableTextareaTabs`: 0 usages
- `Liferay.Util.removeItem`: 0 usages
- `Liferay.Util.resizeTextarea`: 0 usages
- `Liferay.Util.setSelectedValue`: 0 usages
- `Liferay.Util.switchEditor`: 0 usages

Both `Liferay.Util.MAP_HTML_CHARS_ESCAPED` and `Liferay.Util.escapeHTML` have been moved to the `frontend-js-web` module and `deprecated.js` has been marked as deprecated.

### CODE REVIEW

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+const LEFT_SQUARE_BRACKET = '[';
+const RIGHT_SQUARE_BRACKET = ']';
+
+const HTML_ESCAPE = new RegExp(
+	LEFT_SQUARE_BRACKET + htmlUnescapedValues.join('') + RIGHT_SQUARE_BRACKET,
+	'g'
+);
+
+function _escapeHTML(preventDoubleEscape, entities, entitiesValues, match) {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:56]: This function is private (it is not exported from the module), so naming it with a leading underscore is a smell that I think we should avoid. Either it should be inline at the one place it is called (that would be my preferred solution), or it should be given a name of its own (along with a justification for why it can't be inlined — and I can't think of only weak or misguided justifications, personally).

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+export {MAP_HTML_CHARS_ESCAPED};
+
+const MAP_HTML_CHARS_UNESCAPED = {};
+
+const htmlEscapedValues = [];
+
+const htmlUnescapedValues = [];
+
+Object.entries(MAP_HTML_CHARS_ESCAPED).forEach((entry) => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:41]: For readability, instead of accessing `entry[0]` and `entry[1]`, you could use destructuring:

```js
Object.entries(MAP_HTML_CHARS_ESCAPED).forEach(([char, escapedChar]) => {
```

or similar.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	MAP_HTML_CHARS_UNESCAPED[entry[1]] = entry[0];
+
+	htmlEscapedValues.push(entry[1]);
+	htmlUnescapedValues.push(entry[0]);
+});
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:46]: Instead of doing this hand-written slicing and dicing, it seems you could just do:

```js
const htmlEscapedValues = Object.values(MAP_HTML_CHARS_ESCAPED);
const htmlUnescapedValues = Object.keys(MAP_HTML_CHARS_ESCAPED);
```

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:46]: Although as they are "constants", I'd go with all-caps names.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	else {
+		entities = MAP_HTML_CHARS_ESCAPED;
+
+		entitiesValues = htmlEscapedValues;
+	}
+
+	return string.replace(
+		regex,
+		_escapeHTML.bind(null, !!preventDoubleEscape, entities, entitiesValues)
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:114]: This `bind` seems unnecessarily "clever" 🤷‍♂️

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+export default function escapeHTML(string, preventDoubleEscape, entities) {
+	let regex = HTML_ESCAPE;
+
+	const entitiesList = [];
+	let entitiesValues;
+
+	if (entities && typeof entities === 'object') {
+		entitiesValues = [];
+
+		Object.keys(entities).forEach((entry) => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:93]: Another instance where you could use destructuring to improve readability.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+function _escapeHTML(preventDoubleEscape, entities, entitiesValues, match) {
+	let result;
+
+	if (preventDoubleEscape) {
+		const argumentsArray = Array.from(arguments);
+		const length = argumentsArray.length;
+
+		const offset = argumentsArray[length - 2];
+		const string = argumentsArray[length - 1];
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:64]: I find this a bit obfuscated. We know the regex is going to look something like ``/["&'\/<>`]/``, and the call to [`String.prototype.replace` callback function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#Specifying_a_function_as_a_parameter) (ie. this function) is going to get called with the following args, thanks to the `bind` "magic":

```
(
    preventDoubleEscape,
    entities,
    entitiesValues,
    match, // matched substring
    // p1, p2 etc -- capturing subgroups
    offset, // index of match
    string, // whole string being search
)
```

but there will never be any capturing subgroups because of the way we construct the regex pattern (ie. with no parens).

So we could skip all this stuff and change the signature to:

```js
function _escapeHTML(preventDoubleEscape, entities, entitiesValues, match, offset, string) {
```

and just delete these lines:

```js
const argumentsArray = Array.from(arguments);
const length = argumentsArray.length;

const offset = argumentsArray[length - 2];
const string = argumentsArray[length - 1];
```

(Note: in the above I didn't change the name of the function or get rid of the `bind`, but I would do that too.)

I might be tired though, so please double-check this analysis.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+Object.entries(MAP_HTML_CHARS_ESCAPED).forEach((entry) => {
+	MAP_HTML_CHARS_UNESCAPED[entry[1]] = entry[0];
+
+	htmlEscapedValues.push(entry[1]);
+	htmlUnescapedValues.push(entry[0]);
+});
+
+const LEFT_SQUARE_BRACKET = '[';
+const RIGHT_SQUARE_BRACKET = ']';
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/escape_html.js:49]: It seems ridiculous to have named constants for these characters.

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/deprecated.js [hunk 8]
```diff
@@ -14,14 +14,15 @@
 
 // For details about this file see: LPS-2155
 
+/**
+ * @deprecated As of Athanasius (7.3.x), with no direct replacement
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/deprecated.js:18]: I'm not sure it makes sense to deprecate an IIFE. What does that even mean?

We should deprecate the internal APIs that are exposed as side effect of the execution, prior due dilligence on their usage.

**Comment by jbalsas** [comment]
Haven't looked code-wise, but I think @wincent gave plenty feedback there. 

On a more high-level note:
- Let's deprecate the APIs, not arbitrarily executed code
- Feels weird that we expose an `escapeHTML` and not its counterpart `unescapeHTML`...
    - I see usage of both `Lang.String.escapeHTML` and `Lang.String.unescapeHTML`... 
    - Only the `Util.escapeHTML` version is used... 🤔

**Comment by julien** [comment]
> Let's deprecate the APIs, not arbitrarily executed code

Ok that does make more sense

> Feels weird that we expose an escapeHTML and not its counterpart unescapeHTML..

That's true. We'll probably need to expose `unescapeHTML`.

> I see usage of both Lang.String.escapeHTML and Lang.String.unescapeHTML...

Yes those need to be changed too, (correct me if I'm wrong) and that's why we also need the `unescapeHTML` function.

**Comment by julien** [comment]
Here's what I did:

- Added the `unescapeHTML` function in `frontend-js-web`

- Renamed the file `escape_html` to `html_util` because it now exports 2 functions and a constant
  (I know the name might not be the best, but I couldn't come up with something better and I figured 
  that you'd give me feedback on this)

- I replaced the usages of `Lang.String.escapeHTML` and `Lang.String.unescapeHTML` with `Liferay.Util.escapeHTML` and `Liferay.Util.unescapeHTML` respectivelly: 
 I didn't take into account usages that were in `archived` because up to know we've avoided it nor in 
 `frontend-js-aui-web` because those ones will be also moved or deprecated as we work through LPS-112938

- Added `@deprecated` annotation to the unused methods in `deprecated.js`

**Comment by jbalsas** [comment]
> I replaced the usages of Lang.String.escapeHTML and Lang.String.unescapeHTML with Liferay.Util.escapeHTML and Liferay.Util.unescapeHTML

Are we positive these 2 do the same thing?

Ideally, as we endorse an API, we should add some tests to understand what are we actually supporting and what our own API does.

**Comment by wincent** [comment]
> Ideally, as we endorse an API, we should add some tests to understand what are we actually supporting and what our own API does.

Another thing that occurred to me is that, the bigger this change, the more likely we should cc someone from security (I get nervous about anything to do with escaping/unescaping).

**Comment by julien** [comment]
Yes I'm sure `Lang.String.escapeHTML` and `Lang.String.unescapeHTML` are equivalent to `Util.escapeHTML` and `Util.unescapeHTML` respectively.

I also saw that in [1a80d26c74fa2ef41fb447ba2eace252623f47c2](https://github.com/liferay/liferay-portal/commit/1a80d26c74fa2ef41fb447ba2eace252623f47c2), 
the `escapeHTML` function was moved from `portal-web/docroot/html/js/liferay/util.js` to `portal-web/docroot/html/js/liferay/deprecated.js`: this corresponds to [LPS-48464](https://issues.liferay.com/browse/LPS-48464), which says that at the time we were replacing: `Liferay.Util.escapeHTML` to `A.Lang.String.escapeHTML` that is "built into AUI".

I also found this commit (via ` git log -G'\bescapeHTML\b' -p`) [8f40b298866e](https://github.com/liferay/liferay-portal/commit/8f40b298866e) which seems to do the opposite of what this PR is doing, that is going from `Liferay.Util.escapeHTML` to `A.Lang.String.escapeHTML` and `Liferay.Util.unescapeHTML` to `A.Lang.String.unescapeHTML`. As for `escapeHTML` in AUI, I saw that it was added in [afdb7ec4f7](https://github.com/liferay/alloy-ui/commit/afdb7ec4f7)

Concerning test, I'll add those now

**Comment by jbalsas** [comment]
> Yes I'm sure Lang.String.escapeHTML and Lang.String.unescapeHTML are equivalent to Util.escapeHTML and Util.unescapeHTML respectively.

On first inspection, implementations looked pretty different to me... with that in mind, we might want to port over the AUI implementation rather than this one. That would at the very least minimize the risk of breaking stuff to the 3 occurrences of the deprecated usage.

**Comment by julien** [comment]
@jbalsas you're right, there is a difference because `Liferay.Util.escapeHTML` allows 2 additional arguments.

This let you pass a map of "custom" entities so for example you can do something like this:

```javascript
const entities = {
	...Liferay.Util.MAP_HTML_CHARS_ESCAPED,
	...{
		'(': '&#40;',
		')': '&#41;',
		'[': '&#91;',
		']': '&#93;',
	},
};


Liferay.Util.escapeHTML('[()]', false, entities);
// "&#91;&#40;&#41;&#93;"


AUI().Lang.String.escapeHTML('[()]', false, entities);
// "[()]"

``` 

That said, I got confused, because I saw that in [8f40b298866ee53673521c5853af5fcdb76f319f](https://github.com/liferay/liferay-portal/commit/8f40b298866ee53673521c5853af5fcdb76f319f), and more specifically [here](https://github.com/liferay/liferay-portal/commit/8f40b298866ee53673521c5853af5fcdb76f319f#diff-c2d204ddf0fd36e09545d68fa2c2933eR18-R19) it *looked* like we were using both methods in the same way - but if I understand correctly `BBCode.util.escape` receives more arguments that the original `AUI.Lang.String.escapeHTML`. (I don't want to add more confusion, but I wanted to explain what I thought when looking at the different usages) 

Anyway, I'll port the AUI version of these two methods :+1: .

@wincent concerning security, I'm all for it; so maybe we can talk over Slack to see who we can /cc when everything is ready to be merged.

**Comment by wincent** [comment]
> @wincent concerning security, I'm all for it; so maybe we can talk over Slack to see who we can /cc when everything is ready to be merged.

And FWIW, I think async/retrospective review is fine as well, if we can't get a timely review from anybody else (we just released the last GA, so even if we found a problem after merging, we theoretically have time to fast-follow with a fix before the next GA). I'm just flagging this out of caution.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+describe('Liferay.Util.escapeHTML and Liferay.Util.unescapeHTML', () => {
+	describe('Liferay.Util.escapeHTML', () => {
+		it('escapes HTML entities', () => {
+			const escapedEntities = Object.keys(MAP_HTML_CHARS_ESCAPED);
+
+			escapedEntities.forEach((entity) => {
+				expect(escapeHTML(entity)).toEqual(
+					MAP_HTML_CHARS_ESCAPED[entity]
+				);
+			});
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:30]: There is too much indirection in this test. The expected behavior here is hidden inside the `MAP_HTML_CHAR_ESCAPED` constant, which means two undesirable things:

1. I can't tell what the expected behavior by looking at this file.
2. I can break the implementation in ways that keep the test passing.

In tests, "dumb is the new clever". Sticking to DRY (Don't Repeat Yourself) in tests can actually be harmful.

Same concerns apply to the next `describe()`.

The last two `describe()` are better though, because they spell things out a little bit more explicitly. I'll leave a comment below on there though for a suggested improvement.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		});
+	});
+
+	describe('escapeHTML and unescapeHTML used together', () => {
+		it('unescaping an escaped string', () => {
+			const string = "'hello' & <p>goodbye</p>";
+			const escaped = escapeHTML(string);
+			const unescaped = unescapeHTML(escaped);
+
+			expect(unescaped).toEqual(string);
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:52]: This is good because it shows that we can round-trip from unescaped to escaped and back to unescaped, but we could improve it in two ways:

1.  We could actually state that in the description:

    ```js
    describe('roundtripping from unescaped to escaped and back again', () => {
    ```

2.  We could show the intermediate state; eg.

    ```js
    const string = "'hello' & <p>goodbye</p>";

    const escaped = escapeHTML(string);

    expect(escaped).toBe('spell out the literal result here which I am too lazy to type');

    expect(unescaped).toBe(string);
    ```

This is important for the same reason as above: I can break the implementation in ways that make the test continue to "pass", because it isn't specifying the expected behavior of the system as completely as it could.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+		});
+	});
+
+	describe('when using entities non entities', () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:56]: That sentence doesn't parse for me.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js [hunk 12]
```diff
@@ -42,72 +34,20 @@ const HTML_ESCAPED_VALUES = Object.values(MAP_HTML_CHARS_ESCAPED);
...
-
-		regex = new RegExp(entitiesValues.join('|'), 'gi');
-	}
-
-	return string.replace(regex, (match) => entitiesMap[match]);
+export function unescapeHTML(string) {
+	return string.replace(
+		HTML_UNESCAPE,
+		(match) => MAP_HTML_CHARS_UNESCAPED[match]
+	);
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js:52]: This seems quite different to what the old implementation did.

The old one would unescape *any* entity (ie. thanks to [this regex](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L161) and [this fallback](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L404-L410)), whereas your new code will only unescaped these:

```
`&#034;`
`&amp;`
`&#039;`
`&#047;`
`&lt;`
`&gt;`
`&#096;`
```

**Comment by wincent** [comment]
About the commit message on https://github.com/jbalsas/liferay-portal/pull/2193/commits/e94b1fb941d8a2a03963a95cca88b8cf71f7e01e:

> because after checking [e8af3dbeba5413b4c11c826f2cec8b33f054ba89](https://git.io/Jfi9I) and it's corresponding Jira ticket [AUI-1535](https://issues.liferay.com/browse/AUI-1535), I don't think it is necessary.

Can you fill in the gaps here? This states *what* you did (you checked) and *what* you concluded (not needing to carry across from the old code), but doesn't explain *why* you concluded it.

(Don't take this as a criticism — I'm really impressed with how informative your commit messages and PR descriptions are of late; I just want to know what the reasoning is and I would rather have you spell it out for me than me trying to guess and possibly being wrong about it.)

**Comment by julien** [comment]
@wincent yes sure, I can fill the gaps, and there's no offense taken - feedback is always helpful.

I thought that it would be enough to escape the list of entities:

```
&#034;
&amp;
&#039;
&#047;
&lt;
&gt;
&#096;
``` 

But as you [mentioned](https://github.com/jbalsas/liferay-portal/pull/2193#discussion_r433843978), and that was something I wasn't 100% about, it also escapes _any_ entity: so that needs to be done too.

**Comment by wincent** [comment]
Ok. And just to reveal my thought process there, it's the Principle of Least Surprise, which in this case means that I'd want to avoid these kinds of surprises:

1. A function called `unescapeHTML` that operates on some entities (which I can find out about only by looking at the source) but not all.
2. A function that used to work a certain way and no longer does (note that in a very controlled, internal environment, changing the behavior may be totally ok; but when it's public API that may be in use out in the wild in places that we can't see, behavior changes need to be weighed much more carefully).

(For contrast, I *often* write "partial" functions that I know will only handle a subset of the possible inputs, because it ends up being less code, and YAGNI... but those are generally private.)

**Comment by julien** [comment]
There's still one thing I'm not sure about, in theory we should be able to unescape any entity, so I tried with a list from [here](https://www.freeformatter.com/html-entities.html) - not all of them - and they aren't unescaped with `A.Lang.String.unescapeHTML` or (`AUI().Lang.String.unescapeHTML`) 

This is the case with these for example:

```
&#192;
&Ntilde; (and &#209;)
&#180;
```

My intention is to copy `A.Lang.String.unescapeHTML` but I just wanted to mention that.
Anyway, I'll work on fixing this.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js (hunk 11 above)
Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:56]: Yes, that's a mistake. I meant to say "When using strings that look like entities but aren't" - because I actually wanted to cover the errors mentioned in (AUI-1535)[https://issues.liferay.com/browse/AUI-1535]

**Comment by julien** [comment]
I have a reply for my last [comment](https://github.com/jbalsas/liferay-portal/pull/2193#issuecomment-637554215).

I first checked the git logs in the AUI repository and the first commit in which I see `_unescapeHTML` in the `aui-lang.js` file is [9b252b30aa90f3c6b0a167cb30fb156dd9074695](https://github.com/liferay/alloy-ui/commit/9b252b30aa90f3c6b0a167cb30fb156dd9074695),
before that it was in `aui-base.js` but it was identical (at least as far as I can see).  

Later in [e8af3dbeba5413b4c11c826f2cec8b33f054ba89](https://github.com/liferay/alloy-ui/commit/e8af3dbeba5413b4c11c826f2cec8b33f054ba89), it was changed to fix [AUI-1535](https://issues.liferay.com/browse/AUI-1535).

So I think that the reason why this version doesn't unescape other entities is that :

[Here](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L402) we first check to see if the `match` is is in our `MAP_HTML_CHARS_UNESCAPED` and if it isn't then we just use it "as is" and the `if` statement [here](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L404) never evaluates to `true`.
 
Speaking about this `if` statement, there was [something](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L404-L405) that confused me: if `value` is "falsy", why would we call `substr` on it? (ok, ('').substr(1) won't be a problem, but I just found the code confusing).

So I decided to change this as you can see [here](https://github.com/jbalsas/liferay-portal/pull/2193/commits/94f299e0cc260727bac72d5775f9707634d24278), and now we can unescape "more" entities.

The question I have now is:

Would we rather keep an version that doesn't unescape more entities than the ones listed [here](https://github.com/liferay/alloy-ui/blob/835547dd7302c7cd7f5d0c329230f9dd15b93d62/src/aui-base/js/aui-lang.js#L138-L146) to keep the exact same behavior or would be more happy with this one?

And also speaking with Greg who pointed [this](https://stackoverflow.com/questions/1912501/unescape-html-entities-in-javascript/34064434#34064434) out, would that be an option.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 13]
```diff
@@ -13,51 +13,65 @@
...
-					MAP_HTML_CHARS_ESCAPED[entity]
-				);
-			});
+			expect(escapeHTML('"')).toEqual('&#034;');
+			expect(escapeHTML('&')).toEqual('&amp;');
+			expect(escapeHTML("'")).toEqual('&#039;');
+			expect(escapeHTML('/')).toEqual('&#047;');
+			expect(escapeHTML('<')).toEqual('&lt;');
+			expect(escapeHTML('>')).toEqual('&gt;');
+			expect(escapeHTML('`')).toEqual('&#096;');
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:29]: Missing here are tests that prove that it can escape strings containing more than just a single escapable entity (I know your later "round-tripping" test covers that, so it's not a big deal).

My thought process here is: "How could I write broken code that would *still* pass these tests?" and then add tests to catch that kind of breakage. In this case, an example bug would be if the implementation forgot those `g` flags on the regexps — tests like these would still pass.

The basic idea is to always cover representative and edge cases (eg. empty inputs, inputs with things that only partially look like entities, eg. `a & b` etc), go through that exercise of "How could broken code still pass?", write more tests, rinse lather and repeat. You'll never get 100% confidence that you've covered every possible angle, but once you get used to it you can get a long way there very quickly because you develop the mindset that helps you come up with those "representative" and "edge case" examples very quickly. I can't tell you the number of times this process has led me to catch bugs — almost every time I write tests, it seems. 😂

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 14]
```diff
@@ -13,51 +13,65 @@
...
+
 			const unescaped = unescapeHTML(escaped);
 
-			expect(unescaped).toEqual(string);
+			expect(unescaped).toBe(string);
 		});
 	});
 
-	describe('when using entities non entities', () => {
+	describe("when using string that look like entities but aren't", () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:69]: Grammar nit: "string that look like entities" doesn't parse.

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
I looked at the Poshi [report](https://testray.liferay.com/reports/production/logs/test-1-2/1591253128720/test-portal-acceptance-pullrequest(master)/6471/functional-tomcat90-mysql57-jdk8/14/0/LocalFile.WorkflowMetrics_ViewPendingItems/index.html.gz) and from what I understand, test are failing because of something related to "global menu" which makes me think I should rebase this branch master and re-run the tests.

I don't know if it makes sense, so correct me if I'm wrong, but I think it's worth a try.

**Comment by julien** [comment]
ci:test:relevant

**Comment by wincent** [comment]
> I looked at the Poshi [report](https://testray.liferay.com/reports/production/logs/test-1-2/1591253128720/test-portal-acceptance-pullrequest(master)/6471/functional-tomcat90-mysql57-jdk8/14/0/LocalFile.WorkflowMetrics_ViewPendingItems/index.html.gz) and from what I understand, test are failing because of something related to "global menu" which makes me think I should rebase this branch master and re-run the tests.
> 
> I don't know if it makes sense, so correct me if I'm wrong, but I think it's worth a try.

You can certainly try that. In general terms, there is always the `#ci` and `#d-quality-assurance` Slack channels for those cases where you can't make sense of a failure. Jason gave some guidance on where to reach out [here](https://liferay.slack.com/archives/C03BTCQBQ/p1591220732116700), but in short:

- CI infra itself failing (network issues, bot problems etc): report to `#ci`.
- Tests failing: ask QA engineer or post in `#d-quality-assurance`.

If it's just "flaky" or intermittent you can re-run it, but sometimes you need to ask in one of the above places to actually figure out.

**Comment by austinchiang** [comment]
ci:test:relevant

**Comment by jpince** [comment]
ci:test:relevant flaky tests per slack discussion

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js [hunk 15]
```diff
@@ -47,6 +52,18 @@ describe('Liferay.Util.escapeHTML and Liferay.Util.unescapeHTML', () => {
 			expect(unescapeHTML('&#064;')).toEqual('@');
 			expect(unescapeHTML('&#091;')).toEqual('[');
 			expect(unescapeHTML('&#093;')).toEqual(']');
+
+			expect(unescapeHTML('&num;')).toEqual('#');
+			expect(unescapeHTML('&dollar;')).toEqual('$');
+			expect(unescapeHTML('&dollar;')).toEqual('$');
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/html_util.js:58]: This `expect` is duplicated, but given that we've had to try several times in order to get the CI green, I'm inclined to tolerate this until next time we visit this file, rather than running the CI gauntlet again. It's not like it's going to do much harm...

**Comment by wincent** [comment]
ci:forward

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+	'<': '&lt;',
+	'>': '&gt;',
+	'`': '&#096;',
+};
+
+export {MAP_HTML_CHARS_ESCAPED};
+
+const MAP_HTML_CHARS_UNESCAPED = {};
+
+Object.entries(MAP_HTML_CHARS_ESCAPED).forEach(([char, escapedChar]) => {
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js:29]: Hey @julien, I think this might be causing [IE 11 - User cannot sign in](https://issues.liferay.com/browse/LPS-115139). Mind taking a look when you have the chance? 

/cc @wincent

Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js:29]: Yes of course I'm going to look at that

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js:29]: Our core-js polyfill should be setting up `Object.entries` for us. I wonder if there is a race 🤔 — I would have thought the polyfill runs very early on, but maybe not....

Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/html_util.js:29]: That's interesting...

I have sent a fix [here](https://github.com/wincent/liferay-portal/pull/323) seeing that this was an important issue but I can try to see what's happening with the `Object.entries` core-js polyfill on IE11
